crypto/evp/p_lib.c

```diff
@@ -80,66 +81,141 @@ int EVP_PKEY_save_parameters(EVP_PKEY *pkey, int mode)
 
 int EVP_PKEY_copy_parameters(EVP_PKEY *to, const EVP_PKEY *from)
 {
-    if (to->type == EVP_PKEY_NONE) {
-        if (EVP_PKEY_set_type(to, from->type) == 0)
+    int is_provided = 0;
+
+    if (from->ameth == NULL
+        && evp_pkey_make_provided((EVP_PKEY *)from, NULL, NULL, NULL, -1)) {
+        /*
+         * It's fine if the destination hasn't cached anything yet, that
+         * simply makes it a blank page to be filled in with the
+         * evp_keymgmt_copy() call below.
+         * That call will also detect |from->pkeys[0].keymgmt == NULL|,
+         * so we don't bother making an error here.
+         */
+        if (to->pkeys[0].keymgmt != NULL
+            && from->pkeys[0].keymgmt != NULL
+            && (EVP_KEYMGMT_number(to->pkeys[0].keymgmt)
+                != EVP_KEYMGMT_number(from->pkeys[0].keymgmt))) {
+            EVPerr(EVP_F_EVP_PKEY_COPY_PARAMETERS, EVP_R_DIFFERENT_KEY_TYPES);
             return 0;
-    } else if (to->type != from->type) {
-        EVPerr(EVP_F_EVP_PKEY_COPY_PARAMETERS, EVP_R_DIFFERENT_KEY_TYPES);
-        goto err;
+        }
+        is_provided = 1;
+    } else {
+        if (to->type == EVP_PKEY_NONE) {
+            if (EVP_PKEY_set_type(to, from->type) == 0)
+                return 0;
+        } else if (to->type != from->type) {
+            EVPerr(EVP_F_EVP_PKEY_COPY_PARAMETERS, EVP_R_DIFFERENT_KEY_TYPES);
+            return 0;
+        }
     }
 
     if (EVP_PKEY_missing_parameters(from)) {
         EVPerr(EVP_F_EVP_PKEY_COPY_PARAMETERS, EVP_R_MISSING_PARAMETERS);
-        goto err;
+        return 0;
     }
 
     if (!EVP_PKEY_missing_parameters(to)) {
+        /*
+         * TODO(3.0) EVP_PKEY_cmp_parameters() may be inefficient for
+         * comparison of provided domain parameters, can we make it better?
+         */
         if (EVP_PKEY_cmp_parameters(to, from) == 1)
             return 1;
         EVPerr(EVP_F_EVP_PKEY_COPY_PARAMETERS, EVP_R_DIFFERENT_PARAMETERS);
         return 0;
     }
 
-    if (from->ameth && from->ameth->param_copy)
+    /*
+     * If |from| is provided, we ONLY copy the provided bits, and therefore
+     * clear the legacy bits in |to|.
+     */
+    if (is_provided) {
+        evp_pkey_free_legacy(to);
```

Comment:
So a copy parameters operation frees any pre-existing key data? Is that the correct semantics?

Reply:
> frees any pre-existing key data

Just the legacy bits, and only if `from` doesn't have any legacy bits...  but I need to have another look at the `evp_pkey_make_provided` call...